core/src/avm1/globals/date.rs

```diff
@@ -0,0 +1,979 @@
+        .coerce_to_f64(activation)?;
+
+    if new_time.is_finite() {
+        let time = Utc.timestamp_millis(new_time as i64);
+        this.set_date_time(activation.context.gc_context, Some(time));
+        return Ok(time.timestamp_millis().into());
+    }
+
+    this.set_date_time(activation.context.gc_context, None);
+    Ok(NAN.into())
+}
+
+fn set_full_year<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: DateObject<'gc>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let timezone = activation.context.locale.get_timezone();
+    let timestamp = DateAdjustment::new(activation, &timezone)
+        .year(args.get(0))?
+        .month_opt(args.get(1))?
+        .day_opt(args.get(2))?
+        .apply(this);
+    Ok(timestamp.into())
+}
+
+fn set_utc_full_year<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: DateObject<'gc>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let timestamp = DateAdjustment::new(activation, &Utc)
+        .year(args.get(0))?
+        .month_opt(args.get(1))?
+        .day_opt(args.get(2))?
+        .apply(this);
+    Ok(timestamp.into())
+}
+
+pub fn create_date_object<'gc>(
+    gc_context: MutationContext<'gc, '_>,
+    date_proto: Object<'gc>,
+    fn_proto: Option<Object<'gc>>,
+) -> Object<'gc> {
+    let date = FunctionObject::function(
+        gc_context,
+        Executable::Native(constructor),
+        fn_proto,
+        date_proto,
+    );
+    let mut object = date.as_script_object().unwrap();
+
+    object.force_set_function("UTC", create_utc, gc_context, EnumSet::empty(), fn_proto);
+
+    date
+}
+
+pub fn create_proto<'gc>(
+    gc_context: MutationContext<'gc, '_>,
+    proto: Object<'gc>,
+    fn_proto: Object<'gc>,
+) -> Object<'gc> {
+    let date = DateObject::with_date_time(gc_context, Some(proto), None);
+    let mut object = date.as_script_object().unwrap();
+
+    implement_local_getters!(
+        gc_context,
+        object,
+        Some(fn_proto),
+        "getDate" => Datelike::day,
+        "getDay" => |date: &DateTime<FixedOffset>| date.weekday().num_days_from_sunday(),
+        "getFullYear" => Datelike::year,
+        "getHours" => Timelike::hour,
+        "getMilliseconds" => DateTime::timestamp_subsec_millis,
+        "getMinutes" => Timelike::minute,
+        "getMonth" => Datelike::month0,
+        "getSeconds" => Timelike::second,
+        "getYear" => |date: &DateTime<FixedOffset>| date.year() - 1900
+    );
+
+    implement_utc_getters!(
+        gc_context,
+        object,
+        Some(fn_proto),
+        "valueOf" => DateTime::timestamp_millis,
+        "getTime" => DateTime::timestamp_millis,
+        "getUTCDate" => Datelike::day,
+        "getUTCDay" => |date: &DateTime<Utc>| date.weekday().num_days_from_sunday(),
+        "getUTCFullYear" => Datelike::year,
+        "getUTCHours" => Timelike::hour,
+        "getUTCMilliseconds" => DateTime::timestamp_subsec_millis,
```

Comment:
This seems to be the impetus for `impl From<i64> for Value` -- I think it'd be better to make this a closure that explicitly casts to f64 instead of allowing this conversion on Value (which might allow lossy erroneous conversions elsewhere).